LibJS: Implement non standard error.stack attribute

Hendiadyoin1 opened the pull request:
All other browser already support this feature.
There is a Stage 1 proposal to standardize this, but it does not seem
to be active.

linusg reviewed (changes requested):
Please add a test :)

linusg reviewed (commented):
It also appears that other engines put this on the prototype instead, akin to name and message:

![image](https://user-images.githubusercontent.com/19366641/152690621-a949854f-9028-45ab-89e5-9785693bebe4.png)

davidot commented on Userland/Libraries/LibJS/Runtime/Error.h:
```diff
@@ -24,6 +24,7 @@ class Error : public Object {
     virtual ~Error() override = default;
 
     ThrowCompletionOr<void> install_error_cause(Value options);
+    void populate_stack();
```
This function should probably be private or even just not exist and the steps done in the constructor.

Hendiadyoin1 replied:
I made it private for now.

davidot commented on Userland/Libraries/LibJS/Runtime/Error.cpp:
```diff
@@ -48,6 +55,26 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
+    for (size_t i = vm().execution_context_stack().size() - 1; i > 0; --i) {
+        auto const* frame = vm().execution_context_stack()[i];
+        if (auto const* current_node = frame->current_node)
+            stack_string_builder.appendff("{} at {}:{}:{}\n", frame->function_name, current_node->source_range().filename, current_node->source_range().start.line, current_node->source_range().start.column);
```
Although the problem was getting valid `current_node`s there are actually valid execution contexts which don't have an associated scriptormodule and thus no `current_node` so we probably want a check and then just print `func at <unknown>` or something.

davidot reviewed (commented):
Basic code looks fine, just a small couple of comments.
Whether we should add this (it being a stage 1 proposal and all that) I don't know but I don't mind having this if it is widely used on the web.

Hendiadyoin1 commented:
> It also appears that other engines put this on the prototype instead, akin to name and message:

Funnily enough on Chrome/V8 its the other way around:
![grafik](https://user-images.githubusercontent.com/37629766/152700347-09d1fcfc-82e7-404f-a1bc-9061a0483133.png)

linusg commented on Userland/Libraries/LibJS/Tests/error-stack.js:
```diff
@@ -0,0 +1,28 @@
+    let stack_string = (() => {
+        return Error();
+    })().stack;
+    console.log(stack_string);
```
pls don't

linusg commented on Userland/Libraries/LibJS/Tests/error-stack.js:
```diff
@@ -0,0 +1,28 @@
+test("Anonymous function", function () {
+    let stack_string = (() => {
```
We use camelCase in all JS code, to match the built-in functions

Hendiadyoin1 commented:
So I changed the Output style to be more like chromes, this way we include more information
For this I moved the property to the prototype, which turns out is unlike chrome....

Also 2 simple test cases.

linusg commented on Userland/Libraries/LibJS/Tests/error-stack.js:
```diff
@@ -0,0 +1,28 @@
+    })().stack;
+    console.log(stack_string);
+    let [header, ...stack_frames] = stack_string.split("\n");
+    // FIXME: Chrome does not generate a Stack frame for the call to Error()
```
This FIXME probably belongs next to the implementation instead? Also, given that this is non-standard and apparently SM and JSC do different things, we don't need do mirror V8 exactly.

linusg commented on Userland/Libraries/LibJS/Runtime/Error.cpp:
```diff
@@ -48,6 +52,33 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
+{
+    AK::StringBuilder stack_string_builder {};
+
+    // Note: We roughly follow Chromes formatting
```
```suggestion
    // Note: We roughly follow Chrome's formatting
```

linusg commented on Userland/Libraries/LibJS/Runtime/Error.cpp:
```diff
@@ -48,6 +52,33 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
+    AK::StringBuilder stack_string_builder {};
+
+    // Note: We roughly follow Chromes formatting
+    // Note: The errors name and message get prepended by ErrorPrototype::stack
```
```suggestion
    // Note: The error's name and message get prepended by ErrorPrototype::stack
```

linusg commented on Userland/Libraries/LibJS/Runtime/Error.cpp:
```diff
@@ -48,6 +52,33 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
+
+    // Note: We roughly follow Chromes formatting
+    // Note: The errors name and message get prepended by ErrorPrototype::stack
+    // Note: We dont want to capture the global exectution context, so we omit the last frame
```
```suggestion
    // Note: We don't want to capture the global exectution context, so we omit the last frame
```

linusg commented on Userland/Libraries/LibJS/Runtime/Error.cpp:
```diff
@@ -48,6 +52,33 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
+    // Note: We roughly follow Chromes formatting
+    // Note: The errors name and message get prepended by ErrorPrototype::stack
+    // Note: We dont want to capture the global exectution context, so we omit the last frame
+    // FIXME: We seem generate an extra stack-frame when calling an Errors constructor
```
```suggestion
    // FIXME: We seem to generate an extra stack-frame when calling an Errors constructor
```

Well, yes - that's how function calls work. This FIXME is not actionable, please reword it.

Hendiadyoin1 replied:
Is the new wording better?

linusg replied:
Yep :)

linusg commented on Userland/Libraries/LibJS/Tests/error-stack.js:
```diff
@@ -0,0 +1,26 @@
+    let [header, ...stackFrames] = stackString.split("\n");
+
+    expect(header).toBe("Error");
+    expect(!!stackFrames[0].match(/^    at Error \(.*\.js:\d+:\d+\)$/)).toBeTrue();
+    expect(!!stackFrames[1].match(/^    at .*\.js:\d+:\d+$/)).toBeTrue();
```
Can we hardcode the file name (not path) and line numbers at least?

Hendiadyoin1 replied:
I omitted the column numbers, because they seem to point to the end of the word, or the first parenthesis, instead of the beginning of the word, although specifically accessing `source_range.start`

linusg reviewed (approved):
Cool, thanks!
